### PR TITLE
[PBS] add high bitrate sources (HLS), solves issue #3564

### PR DESCRIPTION
**Pull request by chkuendig**

### CODE REVIEW

**Comment by dstftw** [comment]
This fix breaks at least http://www.pbs.org/tpt/constitution-usa-peter-sagal/watch/a-more-perfect-union/ and http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/ for me.

File: youtube_dl/extractor/pbs.py [hunk 1]
```diff
@@ -170,12 +170,36 @@ def _real_extract(self, url):
...
+            
+        formats.append({
+            'format_id': info['recommended_encoding']['eeid'],
+            'url': redirect_info_hls['url'],
+            'protocol': 'm3u8',
+            'ext': 'mp4'
+        })
+        formats.append({
+            'format_id': info['recommended_encoding']['eeid'].replace("hls-800k-16x9", "hls-2500k-16x9"),
+            'url': redirect_info_hls['url'].replace("hls-64-800k", "hls-400-2500k"),
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: What if there is no 2500k quality for the video? Will it work?

**Comment by chkuendig** [comment]
How are they failing? These streams are HLS, so ffmpeg or avconv is required.
Both of your URLs work for me. 
- http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/ redirects to rtmp://sdwfchwqitj2v.cloudfront.net/cfx/st/mp4:videos/newshour/d11d3e6c-abc5-493d-a003-1d3cb623a8b2/112161/hd-mezzanine-16x9/20120223_cyberschool.mp4
- http://www.pbs.org/tpt/constitution-usa-peter-sagal/watch/a-more-perfect-union/ redirects http://ga.video.cdn.pbs.org/videos/constitution-usa-peter-sagal/c8a0021a-9b7e-46d5-9333-72dab59ee845/24476/hd-mezzanine-16x9/CUSP0101_mezz16x9-16x9-hls-400-2500k_930.m3u8

While the download works for both URLs for me (ffmpeg 2.6.1), i can see how the first one might be an issue (its rtmp, but I hardcoded the protocol to m3u8)

Regarding the hardcoded 2500k link: This link is only sent to users logged in on the webpage, so it can't be fetched from youtube-dl and needs to be hardcoded if we want to download it. I didnt't find any video where this hardcoded filename doesnt work. If there is no 2500k quality file it just delivers another video (see the first video above, the redirect_url_hls JSON for 2500k redirects to a 400kbit RTMP stream)

Might it make sense to keep the default URL as before and only add these as additional formats the users can pick from?

If yes, i'll sort them differently (the reference in extractor/common.py said to sort them by quality, but I'm not sure this is the best choice here)

**Comment by dstftw** [comment]
```
python -m youtube_dl http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/ -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.04.09
[debug] Git HEAD: 801a8cf
[debug] Python version 2.7.9 - Windows-2003Server-5.2.3790-SP2
[debug] exe versions: ffmpeg N-68756-g627f565, ffprobe N-68694-g7c210c4, rtmpdump 2.4
[debug] Proxy map: {}
[PBS] education-jan-june12-cyberschools_02-23: Downloading webpage
[PBS] education-jan-june12-cyberschools_02-23: Downloading JSON metadata
[PBS] education-jan-june12-cyberschools_02-23: Downloading alternate_encoding video url info
[PBS] education-jan-june12-cyberschools_02-23: Downloading recommended_encoding video url info
[debug] Invoking downloader on u'rtmp://sdwfchwqitj2v.cloudfront.net/cfx/st/mp4:videos/newshour/d11d3e6c-abc5-493d-a003-1d3cb623a8b2/112161/hd-mezzanine-16x9/20120223_cyberschool.mp4'
[download] Destination: Cyber Schools Gain Popularity, but Quality Questions Persist-2201174722.mp4
ffmpeg version N-68756-g627f565 Copyright (c) 2000-2014 the FFmpeg developers
  built on Dec 27 2014 22:26:33 with gcc 4.9.2 (GCC)
  configuration: --enable-gpl --enable-version3 --disable-w32threads --enable-avisynth --enable-bzlib --enable-fontconfig --enable-frei0r --enable-gnutls --enable-iconv --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libfreetype --enable-libgme --enable-libgsm --enable-libilbc --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-librtmp --enable-libschroedinger --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvo-aacenc --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-decklink --enable-zlib
  libavutil      54. 15.100 / 54. 15.100
  libavcodec     56. 19.100 / 56. 19.100
  libavformat    56. 16.102 / 56. 16.102
  libavdevice    56.  3.100 / 56.  3.100
  libavfilter     5.  6.100 /  5.  6.100
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  1.100 /  1.  1.100
  libpostproc    53.  3.100 / 53.  3.100
Closing connection: NetStream.Play.StreamNotFound
rtmp://sdwfchwqitj2v.cloudfront.net/cfx/st/mp4:videos/newshour/d11d3e6c-abc5-493d-a003-1d3cb623a8b2/112161/hd-mezzanine-16x9/20120223_cyberschool.mp4: Unknown error occurred


ERROR: ffmpeg exited with code 1
  File "c:\Python\Python279\lib\runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\Python\Python279\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Dev\git\youtube-dl\youtube_dl\__main__.py", line 19, in <module>
    youtube_dl.main()
  File "youtube_dl\__init__.py", line 404, in main
    _real_main(argv)
  File "youtube_dl\__init__.py", line 394, in _real_main
    retcode = ydl.download(all_urls)
  File "youtube_dl\YoutubeDL.py", line 1446, in download
    res = self.extract_info(url)
  File "youtube_dl\YoutubeDL.py", line 662, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "youtube_dl\YoutubeDL.py", line 708, in process_ie_result
    return self.process_video_result(ie_result, download=download)
  File "youtube_dl\YoutubeDL.py", line 1150, in process_video_result
    self.process_info(new_info)
  File "youtube_dl\YoutubeDL.py", line 1379, in process_info
    success = dl(filename, info_dict)
  File "youtube_dl\YoutubeDL.py", line 1354, in dl
    return fd.download(name, info)
  File "youtube_dl\downloader\common.py", line 341, in download
    return self.real_download(filename, info_dict)
  File "youtube_dl\downloader\hls.py", line 50, in real_download
    self.report_error('%s exited with code %d' % (ffpp.basename, retval))
  File "youtube_dl\downloader\common.py", line 154, in report_error
    self.ydl.report_error(*args, **kargs)
  File "youtube_dl\YoutubeDL.py", line 529, in report_error
    self.trouble(error_message, tb)
  File "youtube_dl\YoutubeDL.py", line 491, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
```
```
python -m youtube_dl http://www.pbs.org/tpt/constitution-usa-peter-sagal/watch/a-more-perfect-union/ -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'http://www.pbs.org/tpt/constitution-usa-peter-sagal/watch/a-more-perfect-union/', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.04.09
[debug] Git HEAD: 801a8cf
[debug] Python version 2.7.9 - Windows-2003Server-5.2.3790-SP2
[debug] exe versions: ffmpeg N-68756-g627f565, ffprobe N-68694-g7c210c4, rtmpdump 2.4
[debug] Proxy map: {}
[PBS] a-more-perfect-union: Downloading webpage
[PBS] a-more-perfect-union: Downloading player page
[PBS] a-more-perfect-union: Downloading JSON metadata
[PBS] a-more-perfect-union: Downloading alternate_encoding video url info
[PBS] a-more-perfect-union: Downloading recommended_encoding video url info
[debug] Invoking downloader on u'http://ga.video.cdn.pbs.org/videos/constitution-usa-peter-sagal/c8a0021a-9b7e-46d5-9333-72dab59ee845/24476/hd-mezzanine-16x9/CUSP0101_mezz16x9-16x9-hls-400-2500k_930.m3u8'
[download] Destination: A More Perfect Union-2365006249.mp4
ffmpeg version N-68756-g627f565 Copyright (c) 2000-2014 the FFmpeg developers
  built on Dec 27 2014 22:26:33 with gcc 4.9.2 (GCC)
  configuration: --enable-gpl --enable-version3 --disable-w32threads --enable-avisynth --enable-bzlib --enable-fontconfig --enable-frei0r --enable-gnutls --enable-iconv --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libfreetype --enable-libgme --enable-libgsm --enable-libilbc --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-librtmp --enable-libschroedinger --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvo-aacenc --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-decklink --enable-zlib
  libavutil      54. 15.100 / 54. 15.100
  libavcodec     56. 19.100 / 56. 19.100
  libavformat    56. 16.102 / 56. 16.102
  libavdevice    56.  3.100 / 56.  3.100
  libavfilter     5.  6.100 /  5.  6.100
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  1.100 /  1.  1.100
  libpostproc    53.  3.100 / 53.  3.100
[http @ 0000000002c6b5a0] HTTP error 403 Forbidden
http://ga.video.cdn.pbs.org/videos/constitution-usa-peter-sagal/c8a0021a-9b7e-46d5-9333-72dab59ee845/24476/hd-mezzanine-16x9/CUSP0101_mezz16x9-16x9-hls-400-2500k_930.m3u8: Server returned 403 Forbidden (access denied)


ERROR: ffmpeg exited with code 1
  File "c:\Python\Python279\lib\runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "c:\Python\Python279\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Dev\git\youtube-dl\youtube_dl\__main__.py", line 19, in <module>
    youtube_dl.main()
  File "youtube_dl\__init__.py", line 404, in main
    _real_main(argv)
  File "youtube_dl\__init__.py", line 394, in _real_main
    retcode = ydl.download(all_urls)
  File "youtube_dl\YoutubeDL.py", line 1446, in download
    res = self.extract_info(url)
  File "youtube_dl\YoutubeDL.py", line 662, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "youtube_dl\YoutubeDL.py", line 708, in process_ie_result
    return self.process_video_result(ie_result, download=download)
  File "youtube_dl\YoutubeDL.py", line 1150, in process_video_result
    self.process_info(new_info)
  File "youtube_dl\YoutubeDL.py", line 1379, in process_info
    success = dl(filename, info_dict)
  File "youtube_dl\YoutubeDL.py", line 1354, in dl
    return fd.download(name, info)
  File "youtube_dl\downloader\common.py", line 341, in download
    return self.real_download(filename, info_dict)
  File "youtube_dl\downloader\hls.py", line 50, in real_download
    self.report_error('%s exited with code %d' % (ffpp.basename, retval))
  File "youtube_dl\downloader\common.py", line 154, in report_error
    self.ydl.report_error(*args, **kargs)
  File "youtube_dl\YoutubeDL.py", line 529, in report_error
    self.trouble(error_message, tb)
  File "youtube_dl\YoutubeDL.py", line 491, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
```
Both work with old direct links.

For http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/ format codes are the same for each video:
```
python -m youtube_dl http://www.pbs.org/newshour/bb/education-jan-june12-cyberschools_02-23/ -F
[PBS] education-jan-june12-cyberschools_02-23: Downloading webpage
[PBS] education-jan-june12-cyberschools_02-23: Downloading JSON metadata
[PBS] education-jan-june12-cyberschools_02-23: Downloading alternate_encoding video url info
[PBS] education-jan-june12-cyberschools_02-23: Downloading recommended_encoding video url info
[info] Available formats for 2201174722:
format code            extension  resolution note
mirror-mpeg-4-500kbps  mp4        unknown
mirror-mpeg-4-500kbps  mp4        unknown
mirror-mpeg-4-500kbps  mp4        unknown    (best)
```

**Comment by dstftw** [comment]
>Might it make sense to keep the default URL as before and only add these as additional formats the users can pick from?

>If yes, i'll sort them differently (the reference in extractor/common.py said to sort them by quality, but I'm not sure this is the best choice here)

You can use `preference` to force sorting.

**Comment by chkuendig** [comment]
I didn't get it to work with `preference`. In any case, the extractor now doesn't defult to HLS streams anymore and is more conservative to add to the format list in the first place.

Unfortunately i didn't find a way to pull the URL of the high-bitrate stream dynamically (at least not without reverse-engineering the whole delivery-platform PBS built around jwplayer). But for the videos I tested this now allows us to download the HD streams (with `-f hls-2500k-16x9`)

**Pull request closed without merging**